### PR TITLE
Use Browscap-PHP for browser capabilities.

### PR DESCRIPTION
**Pull request by EreMaijala**
Includes util/browscap command line utility for updating the browscap cache.

### CODE REVIEW

File: composer.json
Comment (demiankatz) [inline composer.json]: Might be worth adding this in alphabetical order for consistency.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php]: You say default here, but you use normal below.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php]: Should we register the Guzzle client as a service instead of embedding the logic here? It seems like it would be useful to have a reusable mechanism for obtaining a Guzzle client, and this could be a first small step toward migrating away from the legacy \Laminas\Http client.

**Review by demiankatz — CHANGES_REQUESTED** [review]
See below for a few small suggestions.

The other main concern I have about this is management of cache permissions. If we require a command line tool to update the cache, then users need to set ownership of the browscap cache separately from the rest of the cache. This is certainly not an unsolvable problem, but I could see it being a cause for a lot of support requests and confusion for users less familiar with Unix permissions.

Might it make sense to abstract the browscap update process and hook it not just to the command line client but also to the utilities page of the admin module? That way, it could potentially be updated via the web interface as well, which might be more convenient for some users. If you think this is a good idea but don't have time to implement it, I'm willing to do that refactoring as a follow-up PR once we get the rest of this working.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (EreMaijala) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php]: Fixed. Now "standard" to align with browscap.org definition.

File: composer.json
Comment (EreMaijala) [reply composer.json]: Fixed.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (EreMaijala) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php]: Done. Our proxy mechanism is a bit complicated, but this works.

File: module/VuFind/src/VuFind/Http/GuzzleService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Http/GuzzleService.php]: Is there a mapping for sslcapath? At least in some situations, that setting is required on Debian-based systems.

**Review by demiankatz — CHANGES_REQUESTED** [review]
@EreMaijala, I haven't had a chance to actually test this yet, and I could stand to review some of the code in more depth when time permits... but I spotted one possible issue.

Also, I think you may have overlooked my earlier questions/comments regarding cache permissions (which is kind of buried at the end of my previous review comment -- sorry about the verbosity there!)

File: module/VuFind/src/VuFind/Http/GuzzleService.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Http/GuzzleService.php]: It was missing and sslcafile mapping was invalid. Should be fixed now.

**Comment by EreMaijala** [comment]
> Also, I think you may have overlooked my earlier questions/comments regarding cache permissions (which is kind of buried at the end of my previous review comment -- sorry about the verbosity there!)

I saw it, just didn't manage to get to it yet. :) I'm fine with adding browscap updating to the admin console, but I'm afraid nobody will remember to actually run it if it's a manual task. I suppose that it won't need to be run so often though, so maybe it's a fair tradeoff.

**Comment by EreMaijala** [comment]
@demiankatz Admin Maintenance tool added. I decided to do it separately from the command line utility due to the differences in exception and output handling between the two.

File: module/VuFind/src/VuFind/Http/GuzzleService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Http/GuzzleService.php:198]: I notice that this check is a little more cautious than the one in \VuFindHttp\HttpService (the added initial `$host` check). Do we want to make the same change to the HttpService? If so, do we want to refactor the local check to a trait so it can be reused here? Since this class is (hopefully) just a temporary measure, that may not be necessary or helpful... but if we anticipate the possibility of multiple client-generating services in the long term, it might have some value. In any case, no changes are required here right now regardless of whether we decide to do upstream work.

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/LoginTokenManager.php:232]: Should we do something to help direct the admin to the right place when this happens? Should the exception message include a wiki link, for example? At very least, I think we need to update the [config.ini comment](https://github.com/vufind-org/vufind/blob/dev/config/vufind/config.ini#L603-L608) before merging this.

File: module/VuFindAdmin/src/VuFindAdmin/Controller/MaintenanceController.php
Comment (demiankatz) [inline module/VuFindAdmin/src/VuFindAdmin/Controller/MaintenanceController.php:272]: Not sure if anything can be done about this, but running this with the default memory limit of 128M led to:

`[Thu Feb 08 10:53:45.214654 2024] [php:error] [pid 28947] [client 10.0.2.2:62637] PHP Fatal error:  Allowed memory size of 134217728 bytes exhausted (tried to allocate 4194312 bytes) in /usr/local/vufind/vendor/browscap/browscap-php/src/IniParser/IniParser.php on line 151, referer: http://localhost/vufind_test/Admin/Maintenance`

Raising to 512MB was fine.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php:52]: My concerns about ownership of the browscap cache directory were definitely justified. If the browscap cache directory is not owned by the console user, ALL console utilities are broken by this error:

```
PHP Fatal error:  Uncaught Laminas\Cache\Exception\InvalidArgumentException: Cache directory '/usr/local/vufind/local/cache/browscap' not writable in /usr/local/vufind/vendor/laminas/laminas-cache-storage-adapter-filesystem/src/FilesystemOptions.php:444
Stack trace:
#0 /usr/local/vufind/vendor/laminas/laminas-cache-storage-adapter-filesystem/src/FilesystemOptions.php(126): Laminas\Cache\Storage\Adapter\FilesystemOptions->normalizeCacheDirectory()
#1 /usr/local/vufind/vendor/laminas/laminas-stdlib/src/AbstractOptions.php(118): Laminas\Cache\Storage\Adapter\FilesystemOptions->setCacheDir()
#2 /usr/local/vufind/vendor/laminas/laminas-stdlib/src/AbstractOptions.php(71): Laminas\Stdlib\AbstractOptions->__set()
#3 /usr/local/vufind/vendor/laminas/laminas-cache/src/Storage/Adapter/AdapterOptions.php(347): Laminas\Stdlib\AbstractOptions->setFromArray()
#4 /usr/local/vufind/vendor/laminas/laminas-stdlib/src/AbstractOptions.php(41): Laminas\Cache\Storage\Adapter\AdapterOptions->setFromArray()
#5 /usr/local/vufind/vendor/laminas/laminas-cache-storage-adapter-filesystem/src/FilesystemOptions.php(117): Laminas\Stdlib\AbstractOptions->__construct()
#6 /usr/local/vufind/vendor/laminas/laminas-cache-storage-adapter-filesystem/src/Filesystem.php(132): Laminas\Cache\Storage\Adapter\FilesystemOptions->__construct()
#7 /usr/local/vufind/vendor/laminas/laminas-cache/src/Storage/Adapter/AbstractAdapter.php(82): Laminas\Cache\Storage\Adapter\Filesystem->setOptions()
#8 /usr/local/vufind/vendor/laminas/laminas-cache-storage-adapter-filesystem/src/Filesystem.php(102): Laminas\Cache\Storage\Adapter\AbstractAdapter->__construct()
#9 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/Factory/InvokableFactory.php(25): Laminas\Cache\Storage\Adapter\Filesystem->__construct()
#10 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(620): Laminas\ServiceManager\Factory\InvokableFactory->__invoke()
#11 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(263): Laminas\ServiceManager\ServiceManager->doCreate()
#12 /usr/local/vufind/vendor/laminas/laminas-cache/src/Service/StorageAdapterFactory.php(44): Laminas\ServiceManager\ServiceManager->build()
#13 /usr/local/vufind/vendor/laminas/laminas-cache/src/Service/StorageAdapterFactory.php(39): Laminas\Cache\Service\StorageAdapterFactory->create()
#14 /usr/local/vufind/module/VuFind/src/VuFind/Cache/Manager.php(201): Laminas\Cache\Service\StorageAdapterFactory->createFromArrayConfiguration()
#15 /usr/local/vufind/module/VuFind/src/VuFind/Auth/LoginTokenManagerFactory.php(98): VuFind\Cache\Manager->getCache()
#16 /usr/local/vufind/module/VuFind/src/VuFind/Auth/LoginTokenManagerFactory.php(85): VuFind\Auth\LoginTokenManagerFactory->getBrowscap()
#17 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(620): VuFind\Auth\LoginTokenManagerFactory->__invoke()
#18 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(242): Laminas\ServiceManager\ServiceManager->doCreate()
#19 /usr/local/vufind/module/VuFind/src/VuFind/Auth/ManagerFactory.php(78): Laminas\ServiceManager\ServiceManager->get()
#20 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(620): VuFind\Auth\ManagerFactory->__invoke()
#21 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(242): Laminas\ServiceManager\ServiceManager->doCreate()
#22 /usr/local/vufind/module/VuFind/src/VuFind/Bootstrapper.php(219): Laminas\ServiceManager\ServiceManager->get()
#23 /usr/local/vufind/module/VuFind/src/VuFind/Bootstrapper.php(102): VuFind\Bootstrapper->initUserLanguage()
#24 /usr/local/vufind/module/VuFind/Module.php(86): VuFind\Bootstrapper->bootstrap()
#25 /usr/local/vufind/vendor/laminas/laminas-eventmanager/src/EventManager.php(320): VuFind\Module->onBootstrap()
#26 /usr/local/vufind/vendor/laminas/laminas-eventmanager/src/EventManager.php(170): Laminas\EventManager\EventManager->triggerListeners()
#27 /usr/local/vufind/vendor/laminas/laminas-mvc/src/Application.php(148): Laminas\EventManager\EventManager->triggerEvent()
#28 /usr/local/vufind/vendor/laminas/laminas-mvc/src/Application.php(260): Laminas\Mvc\Application->bootstrap()
#29 /usr/local/vufind/config/application.php(14): Laminas\Mvc\Application::init()
#30 /usr/local/vufind/public/index.php(11): include('...')
#31 {main}

Next Laminas\ServiceManager\Exception\ServiceNotCreatedException: Service with name "Laminas\Cache\Storage\Adapter\Filesystem" could not be created. Reason: Cache directory '/usr/local/vufind/local/cache/browscap' not writable in /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php:627
Stack trace:
#0 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(263): Laminas\ServiceManager\ServiceManager->doCreate()
#1 /usr/local/vufind/vendor/laminas/laminas-cache/src/Service/StorageAdapterFactory.php(44): Laminas\ServiceManager\ServiceManager->build()
#2 /usr/local/vufind/vendor/laminas/laminas-cache/src/Service/StorageAdapterFactory.php(39): Laminas\Cache\Service\StorageAdapterFactory->create()
#3 /usr/local/vufind/module/VuFind/src/VuFind/Cache/Manager.php(201): Laminas\Cache\Service\StorageAdapterFactory->createFromArrayConfiguration()
#4 /usr/local/vufind/module/VuFind/src/VuFind/Auth/LoginTokenManagerFactory.php(98): VuFind\Cache\Manager->getCache()
#5 /usr/local/vufind/module/VuFind/src/VuFind/Auth/LoginTokenManagerFactory.php(85): VuFind\Auth\LoginTokenManagerFactory->getBrowscap()
#6 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(620): VuFind\Auth\LoginTokenManagerFactory->__invoke()
#7 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(242): Laminas\ServiceManager\ServiceManager->doCreate()
#8 /usr/local/vufind/module/VuFind/src/VuFind/Auth/ManagerFactory.php(78): Laminas\ServiceManager\ServiceManager->get()
#9 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(620): VuFind\Auth\ManagerFactory->__invoke()
#10 /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(242): Laminas\ServiceManager\ServiceManager->doCreate()
#11 /usr/local/vufind/module/VuFind/src/VuFind/Bootstrapper.php(219): Laminas\ServiceManager\ServiceManager->get()
#12 /usr/local/vufind/module/VuFind/src/VuFind/Bootstrapper.php(102): VuFind\Bootstrapper->initUserLanguage()
#13 /usr/local/vufind/module/VuFind/Module.php(86): VuFind\Bootstrapper->bootstrap()
#14 /usr/local/vufind/vendor/laminas/laminas-eventmanager/src/EventManager.php(320): VuFind\Module->onBootstrap()
#15 /usr/local/vufind/vendor/laminas/laminas-eventmanager/src/EventManager.php(170): Laminas\EventManager\EventManager->triggerListeners()
#16 /usr/local/vufind/vendor/laminas/laminas-mvc/src/Application.php(148): Laminas\EventManager\EventManager->triggerEvent()
#17 /usr/local/vufind/vendor/laminas/laminas-mvc/src/Application.php(260): Laminas\Mvc\Application->bootstrap()
#18 /usr/local/vufind/config/application.php(14): Laminas\Mvc\Application::init()
#19 /usr/local/vufind/public/index.php(11): include('...')
#20 {main}
  thrown in /usr/local/vufind/vendor/laminas/laminas-servicemanager/src/ServiceManager.php on line 627
```

Must be a factory trying to initialize the cache, since we create all the command objects in the console runner. At very least, we need to find a way to make this more error tolerant and/or to defer the attempt to access the cache.

Comment (demiankatz) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php:52]: On a related note, if the browscap cache directory is owned EXCLUSIVELY by the console user, then the ENTIRE web interface breaks.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala. This seems to be working for the most part, but I think we have a little more documentation work to do, and there are definitely still problems with the cache permissions -- right now, messing up the permissions of the browscap cache directory can have fairly catastrophic consequences. I think we should either find ways to mitigate this, or perhaps defer releasing the console utility until later.

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/LoginTokenManager.php:232]: I'm a bit hesitant on linking to the wiki in an error message, but could be persuaded to. I had updated the comment but somehow lost it. Should be much better now.

File: module/VuFindAdmin/src/VuFindAdmin/Controller/MaintenanceController.php
Comment (EreMaijala) [reply module/VuFindAdmin/src/VuFindAdmin/Controller/MaintenanceController.php:272]: I added `ini_set()` to raise the memory limit automatically for the task.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (EreMaijala) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php:52]: This prompted me to do a bit of work on the cache manager so that initialization of caches is deferred until they're actually needed. I've wanted to do this for a long time since it can save a few cycles for most requests. What do you think? If this is too radical departure from past, I can change it so that other caches are initialized right away, but I'd really like to keep it this way.

What's unfortunate is that Laminas cache requires write access to the directory and its files even if the cache is marked as `['writable' => false]`, read access is not enough.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Now that #3400 is merged, do you mind resolving conflicts here?

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Auth/LoginTokenManager.php:232]: No pressure to link to the wiki in an error message. We can see how well the config.ini setting covers us; if lots of people get confused, we can revisit.

File: module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php
Comment (demiankatz) [reply module/VuFindConsole/src/VuFindConsole/Command/Util/BrowscapCommand.php:52]: Seems reasonable to me, and #3400 is now merged. I'll test all this again after conflicts are resolved, and when I have a little more time (in other words, next week :-) ).

**Comment by EreMaijala** [comment]
@demiankatz Conflicts resolved.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala! Working with the command-line utility is still a bit annoying because of the limitations of Unix permissions, but that's not really our fault, and it's less prone to disastrous consequences than before. I think this is in good shape to merge!

**Pull request merged**
